File: frontend/src/utils/aiBrief.ts
import { differenceInCalendarDays, format, parseISO } from 'date-fns';
import type { Habit, HabitLog } from '../types';
import { calculateDisciplineIndex, calculateDailyCompletion, calculateWeightedScore, getStreak } from './calculations';
import { buildHabitIntentContext } from './habitIntent';

const GROQ_API_URL = 'https://api.groq.com/openai/v1/chat/completions';
const MODEL = 'llama-3.1-8b-instant';
const BRIEF_CACHE_VERSION = 'v2';
const BRIEF_GRACE_DAYS = 3;
const VALID_STATUSES = ['elite', 'solid', 'slipping', 'critical'] as const;
type BriefStatus = (typeof VALID_STATUSES)[number];

export interface BriefOutput {
  status: BriefStatus;
  quote: string;
  motivation: string;
}

interface CachedBrief {
  date: string;
  insight: BriefOutput;
}

const getDaysSinceRegistration = (createdAt?: string): number => {
  if (!createdAt) return 1;
  const parsedDate = parseISO(createdAt);
  if (Number.isNaN(parsedDate.getTime())) return 1;
  return Math.max(1, differenceInCalendarDays(new Date(), parsedDate) + 1);
};

const getStatusFromIndex = (disciplineIndex: number, isGracePeriod: boolean): BriefStatus => {
  if (disciplineIndex >= 80) return 'elite';
  if (disciplineIndex >= 50) return 'solid';
  if (disciplineIndex >= 20) return 'slipping';
  return isGracePeriod ? 'slipping' : 'critical';
};

const clampStatusForGrace = (status: BriefStatus, isGracePeriod: boolean): BriefStatus => {
  if (!isGracePeriod) return status;
  return status === 'critical' ? 'slipping' : status;
};

const trimToWords = (text: string, maxWords: number): string => {
  const words = text.split(/\s+/).filter(Boolean);
  return words.slice(0, maxWords).join(' ');
};

const normalizeQuote = (quote: string): string => {
  const cleaned = quote
    .replace(/[\r\n]+/g, ' ')
    .replace(/\s+/g, ' ')
    .replace(/^["']+|["']+$/g, '')
    .trim();
  const fallback = 'Discipline is built by what you finish today.';
  return trimToWords(cleaned || fallback, 18);
};

const normalizeMotivation = (motivation: string, isGracePeriod: boolean, username: string): string => {
  const defaults = isGracePeriod
    ? [
        `${username}, this is your setup phase. Focus on clean reps today.`,
        'Finish your next one or two protocols and build momentum.',
        'Small consistency now becomes your baseline next week.'
      ]
    : [
        'Execute your next protocol now and raise today\'s score.',
        'Momentum comes from finishing what is still open today.',
        'One clean push now changes tonight\'s result.'
      ];

  const parts = motivation
    .replace(/\r/g, '\n')
    .split(/\n+/)
    .flatMap(line => line.split(/(?<=[.!?])\s+/))
    .map(part => part.replace(/\s+/g, ' ').replace(/^["'\s-]+|["'\s-]+$/g, '').trim())
    .filter(Boolean);

  const lines = parts.slice(0, 3).map(line => trimToWords(line, 16)).filter(Boolean);

  while (lines.length < 2) {
    lines.push(defaults[lines.length]);
  }

  return lines.slice(0, 3).join('\n');
};

const buildFallbackBrief = (status: BriefStatus, isGracePeriod: boolean, username: string): BriefOutput => ({
  status: clampStatusForGrace(status, isGracePeriod),
  quote: isGracePeriod
    ? 'Start simple. Stack one clean win at a time.'
    : 'Your next action decides how this day ends.',
  motivation: normalizeMotivation('', isGracePeriod, username),
});

export async function getDailyBrief(
  username: string,
  habits: Habit[],
  logs: HabitLog[],
  forceRefresh: boolean = false,
  createdAt?: string
): Promise<BriefOutput | null> {
  const apiKey = import.meta.env.VITE_GROQ_API_KEY;
  if (!apiKey) {
    console.error('Groq API key not found in environment.');
    return null;
  }

  const todayFormatStr = format(new Date(), 'yyyy-MM-dd');
  const cacheKey = `ascend_ai_brief_${BRIEF_CACHE_VERSION}_${username}`;
  const historyKey = `ascend_ai_quote_history_${username}`;

  if (!forceRefresh) {
    const cachedStr = localStorage.getItem(cacheKey);
    if (cachedStr) {
      try {
        const cached: CachedBrief = JSON.parse(cachedStr);
        if (cached.date === todayFormatStr) {
          return cached.insight;
        }
      } catch (e) {
        console.error('Error parsing cached brief', e);
      }
    }
  }

  const activeHabits = habits.filter(h => !h.archived);
  const habitIntentContext = buildHabitIntentContext(activeHabits);
  const daysSinceRegistration = getDaysSinceRegistration(createdAt);
  const isGracePeriod = daysSinceRegistration <= BRIEF_GRACE_DAYS;
  const disciplineIndex = calculateDisciplineIndex(habits, logs);
  const todayWeightedScore = Math.round(calculateWeightedScore(habits, logs, todayFormatStr));
  const todayCompletionPercent = Math.round(calculateDailyCompletion(habits, logs, todayFormatStr));
  const targetStatus = getStatusFromIndex(disciplineIndex, isGracePeriod);
  const dayOfWeek = format(new Date(), 'EEEE');

  const habitDetails = activeHabits
    .map(habit => {
      const streak = getStreak(habit.id, logs, todayFormatStr);

      const sevenDaysAgo = new Date();
      sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 6);
      let weeklyCompletion = 0;
      const habitLogs = logs.filter(log => log.habit_id === habit.id && log.status === 'completed');

      for (let i = 0; i < 7; i++) {
        const day = new Date(sevenDaysAgo);
        day.setDate(day.getDate() + i);
        const checkDateStr = day.toISOString().split('T')[0];
        if (habitLogs.some(log => log.date === checkDateStr)) {
          weeklyCompletion++;
        }
      }

      const completedToday = habitLogs.some(log => log.date === todayFormatStr);
      return `- ${habit.name}: ${weeklyCompletion}/7 this week, ${streak} day streak, ${completedToday ? 'done today' : 'not done today'}`;
    })
    .join('\n');

  let quoteHistory: string[] = [];
  try {
    const historyRaw = localStorage.getItem(historyKey);
    if (historyRaw) {
      quoteHistory = JSON.parse(historyRaw);
    }
  } catch (e) {
    console.error('Error reading quote history', e);
  }

  const recentQuotes = quoteHistory.slice(-30);
  const forbiddenQuotesText = recentQuotes.length > 0
    ? `DO NOT use any of these quotes as they were used recently:\n${recentQuotes.map(q => `- "${q}"`).join('\n')}`
    : 'No quotes used recently.';

  const prompt = `You are the AI core of ASCEND. Generate a short Daily Mission Brief.

USER PERFORMANCE DATA:
- Username: ${username}
- Days since registration: ${daysSinceRegistration}
- Grace period active: ${isGracePeriod ? 'YES' : 'NO'} (grace days are days 1-3)
- Discipline Index (DI): ${disciplineIndex}/100
- Daily Weighted Score (DWS, today only): ${todayWeightedScore}/100
- Today's Completion: ${todayCompletionPercent}%
- Day: ${dayOfWeek}
- Protocols:
${habitDetails}
- Habit intent context:
${habitIntentContext}

MATH DEFINITIONS:
- DI = 7-day rolling average of daily weighted scores.
- DWS = single-day weighted score for today only.

TARGET STATUS BASED ON DATA: ${targetStatus}

TONE RULES:
- If grace period is active, be supportive and instructional.
- During grace period, NEVER shame/scold and NEVER return status "critical".
- Outside grace period: elite = appreciate and challenge; solid = direct and steady; slipping = sharp warning; critical = blunt urgency.

Respond ONLY with this exact JSON format:
{
  "status": "elite|solid|slipping|critical",
  "quote": "One short line, max 16 words. NEVER REPEAT A FORBIDDEN QUOTE.",
  "motivation": "2 or 3 short lines separated by \\n. Each line max 16 words. Actionable and easy to scan."
}

Rules:
- Follow the tone rules exactly.
- Use actual habit names from the data.
- Understand each habit using the provided Habit intent context.
- Mention at most 2 habit names total.
- Keep language simple and punchy.
- During grace period, avoid words like "failure", "regret", "pathetic", "waste".
- ${forbiddenQuotesText}

No markdown. No explanation outside the JSON.
IMPORTANT: Only reference habits and data explicitly provided.`;

  try {
    const response = await fetch(GROQ_API_URL, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model: MODEL,
        messages: [{ role: 'user', content: prompt }],
        temperature: 0.9,
        response_format: { type: 'json_object' },
      }),
    });

    if (!response.ok) {
      const err = await response.text();
      console.error('Groq API Error:', err);
      return buildFallbackBrief(targetStatus, isGracePeriod, username);
    }

    const data = await response.json();
    const content = data.choices?.[0]?.message?.content;

    if (!content) {
      return buildFallbackBrief(targetStatus, isGracePeriod, username);
    }

    const parsedRaw = JSON.parse(content) as Partial<BriefOutput>;
    const parsedStatus = VALID_STATUSES.includes(parsedRaw.status as BriefStatus)
      ? (parsedRaw.status as BriefStatus)
      : targetStatus;

    const parsed: BriefOutput = {
      status: clampStatusForGrace(parsedStatus, isGracePeriod),
      quote: normalizeQuote(parsedRaw.quote || ''),
      motivation: normalizeMotivation(parsedRaw.motivation || '', isGracePeriod, username),
    };

    localStorage.setItem(
      cacheKey,
      JSON.stringify({
        date: todayFormatStr,
        insight: parsed,
      })
    );

    quoteHistory.push(parsed.quote);
    localStorage.setItem(historyKey, JSON.stringify(quoteHistory));

    return parsed;
  } catch (error) {
    console.error('Error generating AI Daily Brief:', error);
    return buildFallbackBrief(targetStatus, isGracePeriod, username);
  }
}

File: backend/services/ai_brief.py
"""Port of frontend/src/utils/aiBrief.ts (the "Daily Mission Brief" feature).

Same-day caching and the rolling "don't repeat this quote" history used to
live in localStorage on the client. That stays a frontend concern: the
frontend still owns the cache and passes its recent quote history in on each
call (see BriefRequest.recent_quotes) so the "don't repeat" behavior is
preserved without giving the backend its own persistent store for this.
"""

import json
from datetime import date, datetime, timedelta, timezone
import re
from typing import Optional

from services import calculations, groq_client, habit_intent

VALID_STATUSES = {"elite", "solid", "slipping", "critical"}
BRIEF_GRACE_DAYS = 3


def _get_days_since_registration(created_at: Optional[str]) -> int:
    if not created_at:
        return 1
    try:
        parsed = datetime.fromisoformat(created_at.replace("Z", "+00:00"))
    except ValueError:
        return 1
    today = datetime.now(timezone.utc).date()
    return max(1, (today - parsed.date()).days + 1)


def _get_status_from_index(discipline_index: int, is_grace_period: bool) -> str:
    if discipline_index >= 80:
        return "elite"
    if discipline_index >= 50:
        return "solid"
    if discipline_index >= 20:
        return "slipping"
    return "slipping" if is_grace_period else "critical"


def _clamp_status_for_grace(status: str, is_grace_period: bool) -> str:
    if not is_grace_period:
        return status
    return "slipping" if status == "critical" else status


def _trim_to_words(text: str, max_words: int) -> str:
    words = [w for w in text.split() if w]
    return " ".join(words[:max_words])


def _normalize_quote(quote: str) -> str:
    cleaned = re.sub(r"[\r\n]+", " ", quote)
    cleaned = re.sub(r"\s+", " ", cleaned)
    cleaned = re.sub(r'^["\']+|["\']+$', "", cleaned).strip()
    fallback = "Discipline is built by what you finish today."
    return _trim_to_words(cleaned or fallback, 18)


def _normalize_motivation(motivation: str, is_grace_period: bool, username: str) -> str:
    if is_grace_period:
        defaults = [
            f"{username}, this is your setup phase. Focus on clean reps today.",
            "Finish your next one or two protocols and build momentum.",
            "Small consistency now becomes your baseline next week.",
        ]
    else:
        defaults = [
            "Execute your next protocol now and raise today's score.",
            "Momentum comes from finishing what is still open today.",
            "One clean push now changes tonight's result.",
        ]

    text = motivation.replace("\r", "\n")
    parts = []
    for line in re.split(r"\n+", text):
        for sentence in re.split(r"(?<=[.!?])\s+", line):
            cleaned = re.sub(r"\s+", " ", sentence)
            cleaned = re.sub(r'^["\'\s-]+|["\'\s-]+$', "", cleaned).strip()
            if cleaned:
                parts.append(cleaned)

    lines = [l for l in (_trim_to_words(p, 16) for p in parts[:3]) if l]
    while len(lines) < 2:
        lines.append(defaults[len(lines)])

    return "\n".join(lines[:3])


def _build_fallback_brief(status: str, is_grace_period: bool, username: str) -> dict:
    return {
        "status": _clamp_status_for_grace(status, is_grace_period),
        "quote": "Start simple. Stack one clean win at a time."
        if is_grace_period
        else "Your next action decides how this day ends.",
        "motivation": _normalize_motivation("", is_grace_period, username),
    }


def get_daily_brief(
    username: str,
    habits: list[dict],
    logs: list[dict],
    created_at: Optional[str] = None,
    recent_quotes: Optional[list[str]] = None,
) -> Optional[dict]:
    recent_quotes = recent_quotes or []
    today_str = date.today().isoformat()

    active_habits = [h for h in habits if not h["archived"]]
    habit_intent_context = habit_intent.build_habit_intent_context(active_habits)
    days_since_registration = _get_days_since_registration(created_at)
    is_grace_period = days_since_registration <= BRIEF_GRACE_DAYS
    discipline_index = calculations.calculate_discipline_index(habits, logs)
    today_weighted_score = round(calculations.calculate_weighted_score(habits, logs, today_str))
    today_completion_percent = round(calculations.calculate_daily_completion(habits, logs, today_str))
    target_status = _get_status_from_index(discipline_index, is_grace_period)
    day_of_week = date.today().strftime("%A")

    habit_lines = []
    seven_days_ago = date.today() - timedelta(days=6)
    for habit in active_habits:
        streak = calculations.get_streak(habit["id"], logs, today_str)
        completed_dates = {l["date"] for l in logs if l["habit_id"] == habit["id"] and l["status"] == "completed"}

        weekly_completion = 0
        for i in range(7):
            check_date = (seven_days_ago + timedelta(days=i)).isoformat()
            if check_date in completed_dates:
                weekly_completion += 1

        completed_today = today_str in completed_dates
        habit_lines.append(
            f"- {habit['name']}: {weekly_completion}/7 this week, {streak} day streak, "
            f"{'done today' if completed_today else 'not done today'}"
        )
    habit_details = "\n".join(habit_lines)

    recent = recent_quotes[-30:]
    forbidden_quotes_text = (
        "DO NOT use any of these quotes as they were used recently:\n" + "\n".join(f'- "{q}"' for q in recent)
        if recent
        else "No quotes used recently."
    )

    prompt = f"""You are the AI core of ASCEND. Generate a short Daily Mission Brief.

USER PERFORMANCE DATA:
- Username: {username}
- Days since registration: {days_since_registration}
- Grace period active: {'YES' if is_grace_period else 'NO'} (grace days are days 1-3)
- Discipline Index (DI): {discipline_index}/100
- Daily Weighted Score (DWS, today only): {today_weighted_score}/100
- Today's Completion: {today_completion_percent}%
- Day: {day_of_week}
- Protocols:
{habit_details}
- Habit intent context:
{habit_intent_context}

MATH DEFINITIONS:
- DI = 7-day rolling average of daily weighted scores.
- DWS = single-day weighted score for today only.

TARGET STATUS BASED ON DATA: {target_status}

TONE RULES:
- If grace period is active, be supportive and instructional.
- During grace period, NEVER shame/scold and NEVER return status "critical".
- Outside grace period: elite = appreciate and challenge; solid = direct and steady; slipping = sharp warning; critical = blunt urgency.

Respond ONLY with this exact JSON format:
{{
  "status": "elite|solid|slipping|critical",
  "quote": "One short line, max 16 words. NEVER REPEAT A FORBIDDEN QUOTE.",
  "motivation": "2 or 3 short lines separated by \\n. Each line max 16 words. Actionable and easy to scan."
}}

Rules:
- Follow the tone rules exactly.
- Use actual habit names from the data.
- Understand each habit using the provided Habit intent context.
- Mention at most 2 habit names total.
- Keep language simple and punchy.
- During grace period, avoid words like "failure", "regret", "pathetic", "waste".
- {forbidden_quotes_text}

No markdown. No explanation outside the JSON.
IMPORTANT: Only reference habits and data explicitly provided."""

    raw = groq_client.call_groq(prompt, temperature=0.9, json_mode=True)
    if raw is None:
        return _build_fallback_brief(target_status, is_grace_period, username)

    try:
        parsed_raw = json.loads(raw)
    except (json.JSONDecodeError, TypeError):
        return _build_fallback_brief(target_status, is_grace_period, username)

    parsed_status = parsed_raw.get("status") if parsed_raw.get("status") in VALID_STATUSES else target_status

    return {
        "status": _clamp_status_for_grace(parsed_status, is_grace_period),
        "quote": _normalize_quote(parsed_raw.get("quote") or ""),
        "motivation": _normalize_motivation(parsed_raw.get("motivation") or "", is_grace_period, username),
    }
